qiskit/aqua/operators/primitive_ops/pauli_sum_op.py

```diff
@@ -0,0 +1,405 @@
+            front = StateFn(front, is_measurement=False)
+
+        if isinstance(front, ListOp) and front.distributive:
+            return front.combo_fn(
+                [
+                    self.eval(front.coeff * front_elem)  # type: ignore
+                    for front_elem in front.oplist
+                ]
+            )
+
+        else:
+
+            if self.num_qubits != front.num_qubits:
+                raise ValueError(
+                    "eval does not support operands with differing numbers of qubits, "
+                    "{} and {}, respectively.".format(self.num_qubits, front.num_qubits)
+                )
+
+            if isinstance(front, DictStateFn):
+
+                new_dict = {}  # type: Dict
+                corrected_x_bits = self.primitive.table.X[::-1]  # type: ignore
+                corrected_z_bits = self.primitive.table.Z[::-1]  # type: ignore
+                coeffs = self.primitive.coeffs  # type:ignore
+
+                for bstr, v in front.primitive.items():
+                    bitstr = np.asarray(list(bstr)).astype(np.int).astype(np.bool)
+                    new_b_str = np.logical_xor(bitstr, corrected_x_bits)
+                    new_str = ["".join(map(str, 1 * bs)) for bs in new_b_str]
+                    z_factor = np.product(
+                        1 - 2 * np.logical_and(bitstr, corrected_z_bits), axis=1
+                    )
+                    y_factor = np.product(
+                        np.sqrt(
+                            1
+                            - 2 * np.logical_and(corrected_x_bits, corrected_z_bits)
+                            + 0j
+                        ),
+                        axis=1,
+                    )
+                    for i, n_str in enumerate(new_str):
+                        new_dict[n_str] = (
+                            v * z_factor[i] * y_factor[i] * coeffs[i]
+                        ) + new_dict.get(n_str, 0)
+                    return DictStateFn(new_dict, coeff=self.coeff * front.coeff)
+
+            elif isinstance(front, StateFn) and front.is_measurement:
+                raise ValueError("Operator composed with a measurement is undefined.")
+
+            # Composable types with PauliOp
+            elif isinstance(front, (PauliSumOp, PauliOp, CircuitOp, CircuitStateFn)):
+                return self.compose(front).eval()  # type: ignore
+
+            # Covers VectorStateFn and OperatorStateFn
+            elif isinstance(front, OperatorBase):
+                return self.to_matrix_op().eval(front.to_matrix_op())  # type: ignore
+        return None
+
+    def exp_i(self) -> OperatorBase:
+        """ Return a ``CircuitOp`` equivalent to e^-iH for this operator H. """
+        # TODO: optimize for some special cases
+        from ..evolutions.evolved_op import EvolvedOp
+
+        return EvolvedOp(self)
+
+    def to_instruction(self) -> Instruction:
+        return self.to_matrix_op().to_circuit().to_instruction()  # type: ignore
```

Comment:
Is this type ignore necessary?

Reply:
Yes. If there are no type ignore, 
`qiskit/aqua/operators/primitive_ops/pauli_sum_op.py:311: error: "OperatorBase" has no attribute "to_circuit"`